l2geth/rollup/sync_service_test.go

```diff
@@ -896,14 +899,163 @@ func TestBadFeeThresholds(t *testing.T) {
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tx := mockTx()
+
+	go func() {
+		err = service.applyTransactionToTip(tx)
+	}()
+
+	txLogger.waitForPublish()
+
+	select {
+	case <-txCh:
+		t.Fatal("transaction applied before being logged")
+	default:
+	}
+	txLogger.unblockPublish()
+	<-txCh
+
+	loggedTxMeta := new(QueuedTransactionMeta)
+	buf := <-txLogger.msgs
+	if err := json.Unmarshal(buf, loggedTxMeta); err != nil {
+		t.Fatalf("unable to decode logged transaction meta: %v", err)
+	}
+
+	txMetaJSON, _ := tx.GetMeta().MarshalJSON()
+	loggedTxJSON, _ := json.Marshal(loggedTxMeta)
+	if !bytes.Equal(txMetaJSON, loggedTxJSON) {
+		t.Fatal("mismatched logged transactions")
+	}
+}
+
+func TestSyncServiceTransactionLogFailed(t *testing.T) {
+	txLogger := &failingMockPublisher{}
+	service, txCh, _, err := newTestSyncService(false, nil, txLogger)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tx := mockTx()
+
+	errCh := make(chan error)
+	go func() {
+		err = service.applyTransactionToTip(tx)
+		errCh <- err
+		close(errCh)
+	}()
+
+	if err := <-errCh; err == nil {
+		t.Fatal("transaction applied with log failure")
+	}
+	select {
+	case <-txCh:
+		t.Fatal("transaction applied after log failed")
+	default:
+	}
+}
+
+func TestSyncServiceBackendQueue(t *testing.T) {
+	txLogger := newMockPublisher()
+	txLogger.unblockPublish()
+	activeService, _, _, err := newTestSyncService(false, nil, txLogger)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	queueSub := newMockQueueSubscriber()
+	followerService, followerTxCh, _, err := newTestSyncServiceWithQueueSubscriber(true, nil, &pub.NoopPublisher{}, queueSub)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tx := mockTx()
+	go func() {
+		activeService.applyTransactionToTip(tx)
+	}()
+
+	go func() {
+		followerService.syncTransactionsFromQueue()
+	}()
+
+	// forward the logged transaction from the "active sequencer"
+	msg := <-txLogger.msgs
+	queueSub.ProduceMessage(msg)
+
+	event := <-followerTxCh
+	if len(event.Txs) != 1 {
+		t.Fatal("Unexpected number of transactions")
+	}
+
+	expected, _ := tx.MarshalJSON()
+	eventTx, _ := event.Txs[0].MarshalJSON()
+	if !bytes.Equal(expected, eventTx) {
+		t.Fatalf("invalid tx applied")
+	}
+	expectedMeta, _ := tx.GetMeta().MarshalJSON()
+	eventTxMeta, _ := event.Txs[0].GetMeta().MarshalJSON()
+	if !bytes.Equal(expectedMeta, eventTxMeta) {
+		t.Fatalf("invalid tx meta applied")
+	}
+
+	followerService.chainHeadCh <- core.ChainHeadEvent{}
+	subEvent := <-queueSub.events
+
+	if subEvent != mockAckEvent {
+		t.Fatalf("subscriber failed to acknowledge transaction")
+	}
+}
+
+func TestSyncServiceBackendQueueNack(t *testing.T) {
+	tx := mockTx()
+	tx.GetMeta().RawTransaction = nil
+	tests := map[string][]byte{
+		//"good txmeta":                      []byte("{\"l1BlockNumber\":0,\"l1Timestamp\":1647549225,\"l1MessageSender\":\"0x1487ef4dd5b0ca7610b85964371c1d8ab7c468eb\",\"queueOrigin\":\"sequencer\",\"index\":0,\"queueIndex\":0,\"rawTransaction\":\"34CAgJQrz3UmBr9M0373farCJhgNfaGiVICCAACAgIA=\"}"),
```

Comment:
Does this test fail?

Reply:
No. It's not meant to be tested. The comment is only a reference transaction log I used to construct the other failure scenarios.